Refactoring extended tests

mfojtik opened the pull request:
This PR contains complete refactor of how you write and execute the extended tests. It also provides a new framework you can use to call the `oc` command from within an extended test.

To run the test, you have to have OpenShift server running (with registry and router installed). Then, you can invoke the tests from the `origin` directory:

```console
$ MASTER_ADDR=https://192.168.124.206:8443 SERVER_CONFIG_DIR=/var/lib/openshift/openshift.local.config GOPATH="$(godep path):$GOPATH" go test -v ./test/extended
```

This should produce output like this: https://gist.github.com/mfojtik/65bfe0aa1d23b1441eaa

The new `test/extended/util` package provides a framework that allows you to run any `oc` command. 

For example, following code will process the template and return JSON into `result`:

```go
result, err := oc.Run("process").Args("-f", templatePath).Output()
```

You can also use Go templates to parse the output. In following example, we get the `dockerImageReference` from the ImageStream:

```go
imageName, err := oc.Run("get").Args("is", imageStream).Template("{{ with index .status.tags 0 }}{{ with index .items 0}}{{ .dockerImageReference }}{{ end }}{{ end }}").Output()
```

I think, this will help us write more extended test cases for exotic scenarios and also improves the time of the extended test execution (as you can point it to any OpenShift server as long as you provide admin credentials).

Kudos to @jhadvig for refactoring of `hack/test-extended.sh`

mfojtik commented:
@bparees @smarterclayton @soltysh any early feedback on the framework/cli appreciated :-) We want to add more tests when there will be general agreement that this is the right way.

liggitt commented:
if you already have to have the server running, does this mean we can't test changes that require different server configurations? what configuration does it assume is running?

liggitt commented:
I prefer the approach the integration tests take, where it is trivial to start a server with a default configuration, but each test has a chance to modify the server configuration it will run with. We already use that heavily to test different auth configurations

mfojtik commented:
@liggitt i think the majority of extended test cases will be around testing builds/deployments and I think we will have to start the server in some 'default' configuration to make them consistent. But the current framework can also operate with different servers running, you just have to provide the valid `--config` file or set the configuration for `oc` command to talk to a different server.

liggitt commented:
Benefits to the integration test approach:
* Ability to have the server config contained in the test
* Frees us from side effects of interacting tests by cleaning etcd at the beginning of every test and killing the server at the end of every test
* Allows a single command (`hack/test-integration.sh TestName`) to start the server and run the test

I don't like test suites where I have to manually set up a server first... I find them difficult and inconsistent to run

jhadvig commented:
@liggitt we can also extend functionality that if there is a running OS server launch the extended-tests against it.

mfojtik commented:
@liggitt i think that integration and extended tests have different use-cases.

In integration you want to have different server configs that allows you test things like authentication, use internal REST clients, etc. The problem here is that this is unnecessarily complex. It is difficult to write those tests or read the existing tests. 

Extended tests are more like E2E test, but allows you to test more complex scenarios. In E2E we don't restart the server as well.

liggitt commented:
I vastly prefer the integration tests to the e2e tests

mfojtik commented:
@liggitt writing an integration test for complex deployment with dozen of services will be nightmare.

bparees commented:
@liggitt we also want to avoid starting a server for every TC.  the model we're following is the v2 model.  you can feel free to use randomly generated project/namespaces for each TC to ensure isolation (in fact i think @mfojtik 's framework makes that easy).

if that turns out to be insufficient isolation, we have bigger bugs to fix in openshift itself.

liggitt commented:
I agree about isolation, but I think any extended test framework has to allow testing different server configurations. That implies a server startup and shutdown/cleanup, maybe not per test, but at least for groups of tests.

liggitt commented:
The v2 model made it impossible to test different server configurations... not something I want to repeat

mfojtik commented:
@liggitt a good example is the build_test.go included in this PR... It tests the scenario when you do a STI build and you provide `.sti/environment` file in your source code. We want to verify that the image will be build and the Pod will pickup the environment defined.

Adding this to e2e will make it super slow, because you will have to wait for yet another build to complete. Doing this in integration will on the other hand make it nightmare to write.

mfojtik commented:
@liggitt the framework we have here allows you to point groups of extended tests and target them to specified server with desired configuration (MASTER_ADDR)

liggitt commented:
my point is the target server and desired configuration should be part of the test group / framework so I can run the extended test suite, and the right servers with the right configurations get launched for me

mfojtik commented:
@liggitt the framework here has nothing to do with launching the server. It assumes that the server is running at MASTER_ADDR.

The `./hack/test-extended.sh` launch the server with default config (for hacking). But nothing prevents you to start the server with your configuration and then run the extended tests against it.

jhadvig commented:
Yes only SERVER_CONFIG_DIR and MASTER_ADDRESS. 
Last line in the `test-extended.sh` is launching the tests.

liggitt commented:
hack/test-extended.sh assumes all extended tests will run against a default server configuration, which is unrealistic. The expected server config for a test should be able to be part of the test (or test group) definition somehow.

mfojtik commented:
@liggitt yes, that is good catch. We should probably tag the tests that we want to run against the default server configuration with:

```
# build: extended-default
```

liggitt commented:
@mfojtik so running hack/test-extended.sh wouldn't run all extended tests?

mfojtik commented:
@liggitt yes

mfojtik commented:
@liggitt or we can parametrize test-extended.sh so you will specify which server configuration (test group) you want to run.

liggitt commented:
let's take a step back and make sure we agree on what we want to accomplish here. this is what I would expect from an extended test framework:

1. a command to run all the extended tests that doesn't require a server to be running already
2. a command to run a particular test (or group of tests) that doesn't require a server to be running already
3. the ability to easily write a test (or group of tests) against the default server config
4. the ability to easily write a test (or group of tests) against a non-default server config (in a way that only required the test to tweak the part of the config they wanted to be non-default)
5. the ability to run a group of tests without restarting the server before every one makes sense if:
  * setup is expensive (if particular tests require registry, etc)
  * we're testing isolation by giving each test a namespace and making sure they don't interact

liggitt commented:
I can see the value of letting you run groups of tests against an existing, external, already running server (for testing against old versions, etc), but that doesn't seem like it should be the default, especially if different tests will need different configurations

deads2k commented:
> let's take a step back and make sure we agree on what we want to accomplish here. this is what I would expect from an extended test framework:

We can achieve most of those goals by keeping the framework part of extended-tests to a minimum.  The framework's only responsibility would be to locate the correct (or all) extended-test-bucket.  The extended-text-bucket would have it's own `test.sh` file that could do whatever it wanted.  This would allow individual buckets to decide if and how they wanted to share the cost of server startup.

Different server configurations are essential to getting decent coverage in our extended tests.  Builds and deployments may be able to share a single server using different projects.  They may even be able to run multiple tests in parallel.  However, config tests need to be able to start multiple servers with different starting configurations.  They need to be able to run external kube.   Authentication tests need to be able to spin up external containers to check integration with authentication providers.  Authorization tests may need to destroy all the permissions on a server and attempt to recover them for migration cases. 

Different tests are different and trying to have a restrictive framework won't cover most cases.  We want to enforce a simple entry point and perhaps reporting so that it's easy to locate what's failing, easy to figure out how the code gets executed, and easy to diagnose the problem, but we don't have a need to force everyone into writing tests that are exactly the same shape.  That won't solve our problem.

mfojtik commented:
@deads2k so you're saying we can have something like:

```
test/extended/group1/test.sh
test/extended/group2/test.sh
...
```

where each 'group' will have it's own test.sh that will start the server (and shut it down) and run the test cases from it's folder? The the test-extended.sh just starts all tests sequentially.

I like that :-)

jhadvig commented:
+1

deads2k commented:
> where each 'group' will have it's own test.sh that will start the server (and shut it down) and run the test cases from it's folder? The the test-extended.sh just starts all tests sequentially.

@mfojtik Yes.  I think that keeps the responsibility for server sharing and configuration scoped to a level that allows the individual buckets to be understandable.

smarterclayton commented:
So i want to put the brakes on one particular discussion.

Upstream has an e2e suite.  When we propose things that test like e2e, we
have to first answer why the upstream e2e suite isn't appropriate for our
use.

We have an integration suite (and technology to support that).  When we
propose things that test like integration, we have to first answer why our
integration suite isn't appropriate.

If we need to have separate groups or bundles of tests, we need to answer
why are we reinventing wheels that other people have invented and answer
whether they are truly distinct.

mfojtik commented:
@smarterclayton i checked the upstream e2e framework and I looks promising. We want to test like e2e but not that deep as integration. Extended tests are more high-level feature testing. Things like real deployments with services, all build strategies, secrets in builds, mysql clustering (?), etc... For those things having them as integration will make them very hard to write...

I can see many good things in upstream e2e framework (like waiting for endpoints, getting pods output) which we can re-use here and not reimplement. However, we still need to exercise our CLI when it comes to processing a template and creating resources (and then do build -> deployment).

@jhadvig I think we should now investigate what we can re-use from upstream e2e framework (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/framework.go).

smarterclayton commented:
Well, keep in mind that there are tests that we don't want to always run in
integration.  So "extended" or "group1 / 2" are just names for tests that
we don't run all the time.

On Thu, Jul 16, 2015 at 3:41 PM, Michal Fojtik <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> i checked the
> upstream e2e framework and I looks promising. We want to test like e2e but
> not that deep as integration. Extended tests are more high-level feature
> testing. Things like real deployments with services, all build strategies,
> secrets in builds, mysql clustering (?), etc... For those things having
> them as integration will make them very hard to write...
>
> I can see many good things in upstream e2e framework (like waiting for
> endpoints, getting pods output) which we can re-use here and not
> reimplement. However, we still need to exercise our CLI when it comes to
> processing a template and creating resources (and then do build ->
> deployment).
>
> @jhadvig <https://github.com/jhadvig> I think we should now investigate
> what we can re-use from upstream e2e framework (
> https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/framework.go
> ).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3732#issuecomment-122065511>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton commented:
For exercising the cli, why not have a CLI driven scenario like
hack/test-cmd?  I don't like having lots of shell scripts lying around, but
I don't like reinventing wheels either.

On Thu, Jul 16, 2015 at 3:59 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Well, keep in mind that there are tests that we don't want to always run
> in integration.  So "extended" or "group1 / 2" are just names for tests
> that we don't run all the time.
>
> On Thu, Jul 16, 2015 at 3:41 PM, Michal Fojtik <notifications@github.com>
> wrote:
>
>> @smarterclayton <https://github.com/smarterclayton> i checked the
>> upstream e2e framework and I looks promising. We want to test like e2e but
>> not that deep as integration. Extended tests are more high-level feature
>> testing. Things like real deployments with services, all build strategies,
>> secrets in builds, mysql clustering (?), etc... For those things having
>> them as integration will make them very hard to write...
>>
>> I can see many good things in upstream e2e framework (like waiting for
>> endpoints, getting pods output) which we can re-use here and not
>> reimplement. However, we still need to exercise our CLI when it comes to
>> processing a template and creating resources (and then do build ->
>> deployment).
>>
>> @jhadvig <https://github.com/jhadvig> I think we should now investigate
>> what we can re-use from upstream e2e framework (
>> https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/framework.go
>> ).
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/3732#issuecomment-122065511>.
>>
>
>
>
> --
> Clayton Coleman | Lead Engineer, OpenShift
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

soltysh commented:
My 2 cents: I'd like to see a framework (be it upstream + our addons) that will allow me groups of extended tests, like we do in v2. These usually will be run on demand (all once per day, for example). Additionally I'd like to tag which tests belong to which group, because I want to DRY. The framework should just simplify doing call to OpenShift, exactly like you did that, but on top of that we should be able to test different server configuration (exactly like @liggitt wants) in combination with all different usecases for builds, deployments, etc. IMHO the more tests (different levels covering different aspects) the better for us.
Other than that, good job with that @mfojtik @jhadvig!!!

mfojtik commented:
@smarterclayton I think this https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/examples.go#L59 is something we want to do in extended tests, but using deployments/builds.

smarterclayton commented:
Although if that was bash it would be 5 lines and smaller.

On Jul 17, 2015, at 5:41 AM, Michal Fojtik <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> I think this
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/examples.go#L59
is something we want to do in extended tests, but using deployments/builds.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3732#issuecomment-122230078>.

liggitt commented:
> Although if that was bash it would be 5 lines and smaller.

and completely unmaintainable

deads2k commented:
> and completely unmaintainable

On the other hand, that example is essentially a bash script.  Writing it in Go doesn't help you with detecting changes, since you aren't using proper types for most of it anyway.  Also, the bits that benefit from running as Go (foreachpod: lookForStringInLog), are probably indicative of things that an admin might want to do.  In such a case, it would be good if we were forced to try to make that easier and more understandable.  At worst, it would give an admin an example to follow.

I don't think that's a good example of why we should write our e2e tests in Go.

smarterclayton commented:
I think there's a balance between succinctness of the test (does this
sequence work where i only need to test a few conditions) vs the deep
testing that is possible in the test (I need to verify that all these API
objects change).  The former is why we have hack/test-cmd, the latter is
why you would do a runKubectl() Go function and test.

On Fri, Jul 17, 2015 at 11:06 AM, David Eads <notifications@github.com>
wrote:

> and completely unmaintainable
>
> On the other hand, that example is essentially a bash script. Writing it
> in Go doesn't help you with detecting changes, since you aren't using
> proper types for most of it anyway. Also, the bits that benefit from
> running as Go (foreachpod: lookForStringInLog), are probably indicative of
> things that an admin might want to do. In such a case, it would be good if
> we were forced to try to make that easier and more understandable. At
> worst, it would give an admin an example to follow.
>
> I don't think that's a good example of why we should write our e2e tests
> in Go.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3732#issuecomment-122304159>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

mfojtik commented:
@smarterclayton @deads2k @liggitt I don't think that we are going to replace test-cmd and integration with extended tests (at least not now).

I still think that this PR (and eventually incorporation of the kube e2e framework) have different use cases. I can see a space where we can replace same of the helpers with upstream functions. Also the `CLI` type here is very similar to `kubeCtl` methods.

To summarize this:

1) We are not going to refactor **all** tests (e2e, integration, test-cmd). Only extended tests.
2) Extended tests are meant to test 'extended' scenarios (IOW. cases that take long time to run or are complex and does not fit E2E or integration)
3) We want to be able to group extended tests and allow different configurations of server for a group of tests (@deads2k proposal sounds like it addresses this)

Our goal here is to provide a simple framework, we can re-use in future that will allow us to write the test cases in a simple way, allow to easily debug them for administrators/developers/qe and be sane to maintain (no bash :-)

I'm going to rework the current PR to use 1) gingo to match with upstream, 2) adopt upstream e2e framework where it makes sense, 3) provide two or three example scenarios, 4) refactor to @deads2k proposal to solve grouping.

mfojtik commented:
@smarterclayton @bparees @deads2k @liggitt please take a look on the current "default_mysql_test.go" (https://github.com/mfojtik/origin/blob/extended-tests/test/extended/default_mysql_test.go)

I decided to use a prefix instead of subdirectory for each test to not pollute with a separate Go package for each test. This way it is also easier to share common variables/functions/etc...

mfojtik commented:
@jhadvig FYI ^^

smarterclayton commented:
Looks good

On Mon, Jul 20, 2015 at 1:11 PM, Michal Fojtik <notifications@github.com>
wrote:

> @jhadvig <https://github.com/jhadvig> FYI ^^
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3732#issuecomment-122955761>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

bparees commented:
@mfojtik I miss cucumber but yeah, looks good :)

soltysh commented:
I like it, it's clean and readable :+1:

kargakis commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,221 @@
+}
+
+// Verbose turns on verbose messages for current command
+func (c *CLI) Verbose() *CLI {
```
I really like this pattern of chaining methods (the same is used in the resource builder). :+1:

kargakis commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,221 @@
+	if len(c.Namespace()) == 0 {
+		FatalErr(fmt.Errorf("You must set the namespace before running a command."))
+	}
+	nc := &CLI{
```
Why are you creating and return a new cli struct here? Isn't every method supposed to do it's stuff on the struct and return it instead of returning a new struct?

kargakis replied:
@jhadvig refactor this not to create a new struct but assign everything you need in the struct you already have.

mfojtik replied:
@kargakis there are dragons here... for example if you have two 'Run' in chain, you have to cleanup args (and other fields).

soltysh replied:
I totally agree with @kargakis you break the foo of the [fluent interface](https://en.wikipedia.org/wiki/Fluent_interface). The general approach should be: create new CLI object and perform actions on it, iow.:
```
c = NewCLI().Run("xyz")
```
This way even if somebody by mistake provides two `Run` actions, it'll be run twice, which should be easily to find.

soltysh replied:
This way each method (except for `NewCLI`, but obviously that will be the only one not part of `CLI` structure) will modify only the contents of `CLI` structure.

mfojtik replied:
@soltysh we are going to have two, NewCLI and NewAdminCLI, etc... what function you use then? My expectation would be that the .Run() method should re-use authentication and config from the first initialization when chained...

```go
oc := NewCLI()
oc.Run("create").Run("delete")
```

Both 'create' and 'delete' should use the same authentication setup... Your suggestion is making this a bit harder ;-)

soltysh replied:
Why, you'll be setting up the `CLI` structure differently inside appropriate `New...` method. Then the `Run` and others method (if needed) will be responsible for using appropriate set of values from the `CLI` struct. I don't see any impediment in doing that this way, but maybe because I do have some Java background :stuck_out_tongue_winking_eye: and was doing exactly that a good couple of time

kargakis replied:
> fluent interface

Ah, so that's how it's called:)

soltysh replied:
Yeah, I've googled that, I've always called it a builder pattern :wink:  but the latter gave me that name actually.

jhadvig commented:
@mfojtik I've finished with extended-tests for sti-build. Also did a bit of refact in the framework.
@bparees @soltysh @kargakis  @smarterclayton PTAL

kargakis commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,225 @@
+	}
+	client.adminConfigPath = configPath
+	if runInGinko {
+		func() { BeforeEach(client.beforeEach) }()
```
Why are you running this in a closure?

mfojtik replied:
to allow to use this framework outside Gingo :-) Which also invalidates your comment below. @deads2k suggested (and I agree with that) that requiring Ginko in place is a hidden requirement and disallows to use this outside Ginko.

Also 'BeforeEach' is dangerous because it installs the hook on behalf without telling you. Having this as argument makes you accept your fate :-)

kargakis commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,225 @@
+}
+
+// SetNamespace overrides the namespace set by framework
+func (c *CLI) SetNamespace(ns string) {
```
Make this
```go
func (c *CLI) SetNamespace(ns string) *CLI {
```

and above in `NewCLI`:
```go
if !runInGinko {
   return client.SetNamespace(project)
}
BeforeEach(client.beforeEach)
return client
```
Though I am not sure about the closure around `BeforeEach` (see comment above).

kargakis commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,225 @@
+		Logf("Adding project role bindings to %q namespace", c.Namespace())
+		c.SetupRoleBindings()
+	} else {
+		Failf("Framework does not have the namespace set")
```
Is this `Failf` exiting the tests?

mfojtik replied:
just the actual test case.

kargakis commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,225 @@
+	client, err := testutil.GetClusterAdminClient(c.adminConfigPath)
+	if err != nil {
+		FatalErr(err)
+		return nil
```
I see that `FatalErr` is calling `os.Exit` so you don't need this return.

kargakis commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,225 @@
+	client, err := testutil.GetClusterAdminKubeClient(c.adminConfigPath)
+	if err != nil {
+		FatalErr(err)
+		return nil
```
Same as above, you don't need this return.

kargakis commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,225 @@
+
+// Namespace returns the name of the namespace used in the current test case
+func (c *CLI) Namespace() string {
+	if c.kubeFramework == nil {
```
```go
if c.kubeFramework == nil || c.kubeFramework.Namespace == nil {
```

kargakis commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,225 @@
+func (c *CLI) Template(t string) *CLI {
+	if c.verb != "get" {
+		FatalErr("Cannot use Template() for non-get verbs.")
+		return c
```
No need for this return

kargakis commented on test/extended/util/framework.go:
```diff
@@ -0,0 +1,115 @@
+
+		w, err := f.Client.Builds(f.Namespace.Name).Watch(
+			labels.Everything(),
+			fields.Set{"name": buildName}.AsSelector(),
```
Is "name" working? https://github.com/openshift/origin/blob/34260915f85b9520c4ab17f8c7ef648718878588/pkg/build/registry/build/strategy.go#L108

mfojtik replied:
@kargakis I don't think this selectors are working for builds (yet).

jhadvig replied:
Yep they are working, I was a bit confused too, why the `metadata.name` is not working and `name` does.

kargakis commented on test/extended/util/framework.go:
```diff
@@ -0,0 +1,115 @@
+
+// WaitForABuild waits for a build to become Completed
+func (f *OsFramework) WaitForABuild(buildName string) error {
+	for {
```
Do you need this for and all the code (except `isOK` and `isFailed`) up to the watcher?

mfojtik replied:
this is copied from upstream e2e

deads2k replied:
I don't see any timeout on this.  Is there some global timeout that will fail this eventually?

deads2k commented on test/extended/default_test.sh:
```diff
@@ -0,0 +1,221 @@
+	SERVER_HOSTNAME_LIST="${SERVER_HOSTNAME_LIST},${IP_ADDRESS}"
+done <<< "${ALL_IP_ADDRESSES}"
+
+openshift admin ca create-master-certs \
```
This looks very similar to `hack/test-end-to-end.sh` and `hack/test-cmd.sh`.  Is there a clean way to factor it out?

deads2k commented on test/extended/default_test.sh:
```diff
@@ -0,0 +1,221 @@
+echo "[INFO] MASTER IP - ${MASTER_ADDR}"
+echo "[INFO] SERVER CONFIG PATH - ${SERVER_CONFIG_DIR}"
+
+MASTER_ADDR="${MASTER_ADDR}" SERVER_CONFIG_DIR="${SERVER_CONFIG_DIR}" GOPATH="$(godep path):/${GOPATH}" go test -v -tags=default ./test/extended
```
Rather than pass a `MASTER_ADDR`, how about setting `KUBECONFIG` to the `admin.kubeconfig` file?

soltysh replied:
+1

deads2k commented on test/extended/extended_test.go:
```diff
@@ -0,0 +1,70 @@
+package extended
+
+import (
+	"flag"
```
Why aren't we using `spf13/pflag` like our main code path?

mfojtik replied:
This is copying the E2E framework.

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/e2e_test.go#L20

deads2k replied:
> This is copying the E2E framework.

Copying or not, why are using a different "flag" library?  Any reason besides cargo-culting?

deads2k commented on test/extended/extended_test.go:
```diff
@@ -0,0 +1,70 @@
+	config.GinkgoConfig.RandomizeAllSpecs = false
+
+	flag.StringVar(&testContext.KubeConfig, clientcmd.RecommendedConfigPathFlag, adminKubeConfigPath(), "Path to kubeconfig containing embeded authinfo.")
+	flag.StringVar(&testContext.CertDir, "cert-dir", kubeCertPath(), "Path to the directory containing the certs. Default is empty, which doesn't use certs.")
```
What is this used for?

deads2k commented on test/extended/extended_test.go:
```diff
@@ -0,0 +1,70 @@
+// init initialize the extended testing suite.
+// You can set these environment variables to configure extended tests:
+// MASTER_ADDR - The host or apiserver to connect to
+// SERVER_CONFIG_DIR - Directory where OpenShift stores configuration
```
seems like if you want the config, point directly to the `master-config.yaml`

deads2k commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,225 @@
+// If runInGinko argument is set to true, this function will install BeforeEach
+// hook to Gingo test case that grant project role bindings to a namespace created
+// by the E2E test. This option must be set inside Describe() or Context()
+func NewCLI(project, configPath string, runInGinko bool) *CLI {
```
Passing a bool here is pretty kludgy.  Is there any way to auto-detect?  If not, I'd prefer two public methods `NewGinkoCLI` and `NewCLI` to make the intent more clear.

mfojtik replied:
Make sense. @jhadvig can you change this?

deads2k commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,225 @@
+// REST clients.
+type CLI struct {
+	verb            string
+	adminConfigPath string
```
This really requires cluster-admin rights to operate?

mfojtik replied:
no, this can be renamed to 'configPath'

deads2k commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,225 @@
+// to a buffer and prepare the global flags such as namespace and config path.
+func (c *CLI) Run(verb string) *CLI {
+	out := new(bytes.Buffer)
+	if len(c.Namespace()) == 0 {
```
This seems problematic.  How would I manipulate cluster-scoped resources like `users`?

mfojtik replied:
@deads2k I think we can relax on this check and make the command fail on it's own when the namespace is not set and it has to be set.

soltysh replied:
I was reading and thought about the same think @deads2k mentioned, how about `RunAsClusterAdmin` method to test cluster-scoped resources?

deads2k replied:
> I was reading and thought about the same think @deads2k mentioned, how about RunAsClusterAdmin method to test cluster-scoped resources?

We don't always want to run it as a cluster admin.  Relaxing the check and letting commands fail on their own will cover all the cases.

mfojtik commented on test/extended/util/framework.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+// CreatePodObjectForImageStream creates a pod object from given imageStream
+func (f *OsFramework) CreatePodObjectForImageStream(imageStreamName string) (*kapi.Pod, error) {
```
`CreatePodFromImageStream`

bparees replied:
CreatePodForImageStream is shorter still. :)

mfojtik commented on test/extended/util/framework.go:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	imageName := imageStream.Status.Tags[tags[0]].Items[0].DockerImageReference
+	podName := "test-pod" + string(kutil.NewUUID())
```
use util.Namer package for this

deads2k commented:
I think the bash scripts that run the extended tests should exist outside of a go package.  Keeping them in a separate directory makes for a better logical division.

jhadvig commented:
Well the idea here was to have not one bash script, but several, which will simulate different environment setups, thats why we put the `default_test` script into the test/extended dir, cause more of them will come over time.

deads2k commented:
> Well the idea here was to have not one bash script, but several, which will simulate different environment setups, thats why we put the default_test script into the test/extended dir, cause more of them will come over time.

I like having multiple.  I probably even like having separate directories for each one, but I'm not crazy about having shell scripts in my go package.  Maybe under `hack/test-extended/`?

ironcladlou commented:
>>Well the idea here was to have not one bash script, but several, which will simulate different environment setups, thats why we put the default_test script into the test/extended dir, cause more of them will come over time.

>I like having multiple. I probably even like having separate directories for each one, but I'm not crazy about having shell scripts in my go package. Maybe under hack/test-extended/?

+ to `hack/test-extended` for the scripts.

mfojtik commented:
@deads2k separate directories will also require having separate go packages for each group, which will complicate imports (and package namings...).... I like ./hack/test-extended/{default,foo,bar}.sh ?

jhadvig commented:
+1 for the ./hack/test-extended/{default,foo,bar}.sh

bparees commented on test/extended/util/framework.go:
```diff
@@ -0,0 +1,115 @@
+		tags = append(tags, tag)
+	}
+
+	imageName := imageStream.Status.Tags[tags[0]].Items[0].DockerImageReference
```
might be no tags?

bparees replied:
also picking the first tag feels a little arbitrary.  should we punt this method and just have a CreatePodForImageStreamTag instead, where you have to be explicit about the tag you want?

mfojtik commented:
@deads2k @bparees @jhadvig @kargakis all comments (hopefully) addressed, can you please have second look? Summary of changes:

* Squashed/logically re-organized all commits
* Clean up the startup options. You have to set `MASTER_ADDR` and `KUBECONFIG` env vars when running against existing server.
* Moved the starter script to `./hack/test-extended/default.sh` (this will start the "default_" group
* Added `oc exec` verification to STI environment build
* SetNamespace() is now chained
* Added NewGingoCLI (to get rid of 'runInGingo' option)

mfojtik commented:
[test]

mfojtik commented:
@smarterclayton can you please review this?

smarterclayton commented:
It's in my queue but may not get to it until monday

smarterclayton commented:
I don't have any issues with where you guys are right now.

@liggitt any final thoughts about this structure?

liggitt commented:
> @liggitt any final thoughts about this structure?

I'm not a huge fan of the ginkgo stuff, but don't have a different implementation to recommend at the moment. @deads2k has looked at this more than I have... if he's ok, I'm ok.

deads2k commented on hack/test-extended/default.sh:
```diff
@@ -0,0 +1,224 @@
+	SERVER_HOSTNAME_LIST="${SERVER_HOSTNAME_LIST},${IP_ADDRESS}"
+done <<< "${ALL_IP_ADDRESSES}"
+
+openshift admin ca create-master-certs \
```
This block is common (or so similar I don't see the difference right off) to here, `test-cmd.sh`, and `test-end-to-end.sh`.  Can you refactor to dry it out?

deads2k replied:
And this block?  Server config building followed by startup?

deads2k commented on hack/test-extended/default.sh:
```diff
@@ -0,0 +1,224 @@
+# set path so OpenShift is available
+export PATH="${GO_OUT}:${PATH}"
+
+cleanup() {
```
another common block between our test scripts.

jhadvig replied:
If you take a closer look you will see that the methods are not so familiar. Al the `cleanup()` in the ext-tests is doing is stops OS server, stops and removes containers, removes the testing dir (+ unmounts any mounts used)

deads2k commented on hack/test-extended/default.sh:
```diff
@@ -0,0 +1,224 @@
+fi
+
+
+wait_for_url "${KUBELET_SCHEME}://${KUBELET_HOST}:${KUBELET_PORT}/healthz" "[INFO] kubelet: " 0.5 60
```
another common block with other test scripts.

deads2k commented on hack/test-extended/default.sh:
```diff
@@ -0,0 +1,224 @@
+source ${OS_ROOT}/hack/util.sh
+source ${OS_ROOT}/hack/common.sh
+
+echo "[INFO] Starting extended tests"
```
print out which bucket.  "default" in this case

deads2k commented on hack/test-extended/default.sh:
```diff
@@ -0,0 +1,224 @@
+echo "[INFO] MASTER IP - ${MASTER_ADDR}"
+echo "[INFO] SERVER CONFIG PATH - ${SERVER_CONFIG_DIR}"
+
+MASTER_ADDR="${MASTER_ADDR}" \
```
`MASTER_ADDR` is a subset of information available in the `KUBECONFIG`.  I don't think you need this.

deads2k commented:
Given multiple buckets, proper factorization inside the `test/extended` package is going to be important.  Go ahead and separate the tests into packages based on area.  It might be a pain now, but you'll thank yourself later when we've got half a dozen test areas.

deads2k commented on test/extended/util/framework.go:
```diff
@@ -0,0 +1,110 @@
+			return e.Name == buildName && e.Status.Phase == buildapi.BuildPhaseComplete
+		}
+
+		isFailed := func(e *buildapi.Build) bool {
```
shouldn't this be limitted by build name too?

deads2k commented on test/extended/util/framework.go:
```diff
@@ -0,0 +1,110 @@
+// WaitForABuild waits for a build to become Completed
+func (f *OsFramework) WaitForABuild(buildName string) error {
+	for {
+		list, err := f.Client.Builds(f.Namespace.Name).List(labels.Everything(), fields.Everything())
```
Why not have the same field selector as the watch below?

deads2k commented on test/extended/util/framework.go:
```diff
@@ -0,0 +1,110 @@
+
+// OsFramework provides helper functions and watchers for OpenShift
+type OsFramework struct {
+	Namespace *kapi.Namespace
```
Why do we want to have one these for each namespace?  Also, what value does this bundling have for me?  Why wouldn't I simply have a `WaitForABuild` helper function that took a `BuildInterface` and a `buildName`?

mfojtik replied:
I was trying to make this consistent with kube framework WaitFor..(..). I agree that having this as a standalone function with BuildInterface, buildName and namespace will be much nicer

deads2k commented on test/extended/default_builds_test.go:
```diff
@@ -0,0 +1,80 @@
+	Describe("Building from a template", func() {
+
+		It(fmt.Sprintf("should create a image from %q template and run it in a pod", stiEnvBuildFixture), func() {
+			outputPath := getTempFilePath(oc.Namespace())
```
move this closer to point of use.

mfojtik replied:
I removed this function.

deads2k commented on test/extended/util.go:
```diff
@@ -0,0 +1,23 @@
+	return os.Getenv("KUBECONFIG")
+}
+
+func writeTempJSON(path, content string) error {
```
Seems weird to have a `WriteTempJSON` that doesn't have to write to a temporary file.  Why not have it choose a temp file and return the path it wrote?

deads2k replied:
Accepting a `runtime.Object` will make this utility easier to use.  As it is, there's no reason to expect JSON to be delivered to this method.

mfojtik replied:
@deads2k there are two use-cases for this... one is storing a runtime.Object to a file and another is storing a string (basically an output from oc...Output()). I added two functions:

writeObjectToFile()
writeToFile()

(the first one does the decoding of runtime.Object and just call the second one at the end.)

deads2k commented on test/extended/util/framework.go:
```diff
@@ -0,0 +1,110 @@
+
+// CreatePodForImageStreamTag creates a pod object from given imageStream and the tag
+func (f *OsFramework) CreatePodForImageStreamTag(imageStreamName, tag string) (*kapi.Pod, error) {
+	imageStream, err := f.Client.ImageStreams(f.Namespace.Name).Get(imageStreamName)
```
This get isn't logically required.  Why not help create the `Pod`, but leave it up to the caller to decide how much checking he wants.  Maybe he wants to create a `Pod` that will have a "cannot pull" event on it.

mfojtik replied:
Yes, I've split this into two functions (GetDockerImageReference() and CreatePodForImage())

deads2k commented on test/extended/util/framework.go:
```diff
@@ -0,0 +1,110 @@
+			Labels: map[string]string{"name": podName},
+		},
+		Spec: kapi.PodSpec{
+			ServiceAccountName: "builder",
```
This should either be a parameter or left blank.

deads2k commented on test/extended/default_builds_test.go:
```diff
@@ -0,0 +1,80 @@
+
+	var imageStreamFixture = filepath.Join("..", "integration", "fixtures", "test-image-stream.json")
+	var stiEnvBuildFixture = filepath.Join("fixtures", "test-env-build.json")
+	var oc = exutil.NewGinkoCLI("build-sti-env", kubeConfigPath())
```
This test should not run as a `cluster-admin`.  You can use your `cluster-admin` rights to create a new project, assign an admin, create an editor, and then run the rest of the tests as the editor.

deads2k replied:
Thinking on it, you should actually login in as a non-privileged user, request a new user, and then make certain that the default rights created in the new project are sufficient to run this test.  If they aren't, we should either fix the defaults or describe in this test the code required to get those rights.

deads2k commented on test/extended/default_builds_test.go:
```diff
@@ -0,0 +1,80 @@
+			}
+
+			By("expecting the pod to be running")
+			Expect(oc.KubeFramework().WaitForPodRunning(pod.Name)).NotTo(HaveOccurred())
```
This reads really strangely as a sentence. I think the intent is "start this pod and expect no errors".  As written, the sentence reads as "I expect wait for pod running to have failed.".

mfojtik replied:
Yeah, I will store the error into 'err' and then just do Expect(err) which reads more clearly

deads2k commented on test/extended/default_builds_test.go:
```diff
@@ -0,0 +1,80 @@
+
+			By(fmt.Sprintf("calling oc create -f %q", outputPath))
+			if err := oc.Run("create").Args("-f", outputPath).Verbose().Execute(); err != nil {
+				Failf("Unable to create pod: %v", err)
```
what aren't these, `Expect(err).NotTo(HaveOccurred())`?

mfojtik replied:
Yes, I also fixed this in other test cases

deads2k commented on test/extended/extended_test.go:
```diff
@@ -0,0 +1,71 @@
+
+// init initialize the extended testing suite.
+// You can set these environment variables to configure extended tests:
+// MASTER_ADDR - The host, or apiserver to connect to
```
This should be unnecessary if we require a `KUBECONFIG` which seems more appropriate.

deads2k commented on test/extended/default_mysql_test.go:
```diff
@@ -0,0 +1,49 @@
+// +build default
```
Have you tested to be sure that these work how we expect them to when using ginkgo?

mfojtik replied:
If I change this some something else than default, the go file is being ignored when building, so the test case is not executed.

deads2k commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,219 @@
+}
+
+// AdminRESTClient return the current project namespace REST client
+func (c *CLI) AdminRESTClient() *client.Client {
```
I'm pretty sure this isn't guaranteed to be a client for an admin.

deads2k commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,219 @@
+
+// Namespace returns the name of the namespace used in the current test case
+func (c *CLI) Namespace() string {
+	if c.kubeFramework == nil || c.kubeFramework.Namespace == nil {
```
You don't make this nil check for `kubeFramework` elsewhere.  If we require it, force it during construction.  Otherwise, check it everywhere.  Probably easier to require it.

mfojtik replied:
Right, the kubeFramework is set in both initializers, so this check doesn't make sense at all ;-) I will just remove it.

deads2k commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,219 @@
+// Execute executes the current command and return error if the execution failed
+// This function will set the default output to stdout.
+func (c *CLI) Execute() error {
+	out, err := c.Output()
```
This first err is ignored.

mfojtik replied:
no longer

deads2k commented on test/extended/util/cli.go:
```diff
@@ -0,0 +1,219 @@
+	}
+	out := c.stdout.(*bytes.Buffer)
+
+	return strings.TrimSpace(out.String()), err
```
I would have expected verbose output to include the output of the commands.

deads2k commented:
I see two blockers:
 1. Using cluster-admin to run tests like this is bad practice because it doesn't ensure that the scenario being tested will actually work for users.
 1. Building a config and launching a default openshift from a bash script needs to be factored out for consistency.

deads2k commented:
Commit cleanup.

When you squash down, be sure to update your commit names for your bumps.  It should be "bump(github.com/onsi/ginkgo) d981d36e9884231afa909627b9c275e4ba678f90" and "bump(github.com/GoogleCloudPlatform/kubernetes/test/e2e) cd821444dcf3e1e237b5f3579721440624c9c4fa".

jhadvig commented:
@deads2k I've already submitted PR to @mfojtik branch, for dry-ing our test-* scripts -> https://github.com/mfojtik/origin/pull/7

mfojtik commented:
@deads2k PTAL

deads2k commented:
Alright, squash down where it makes sense, merge in mfojtik#7, update the UPSTREAM pull with PR number and I think this is good to go.

mfojtik commented:
@deads2k all done (upstream PR: https://github.com/GoogleCloudPlatform/kubernetes/pull/12221)

Also I added one more helper function (oc.Run().OutputToFile(name)) to get rid if WriteToFile and make the tests look more cleaner.

deads2k commented:
@mfojtik rename the commit to "UPSTREAM: 12221: Allow custom namespace creation in e2e framework" to align with our pattern.

Then lgtm and feel free to merge.

jhadvig commented:
@mfojtik @deads2k ok so the test-cmd is finally working. Had to refactor a bit but its working now.
I guess its good time to merge this.

deads2k commented:
[merge]

mfojtik commented:
re-[merge]

jhadvig commented:
@mfojtik :joy:

deads2k commented:
@jhadvig The party isn't over yet.  Have you guys gotten up with @danmcp to see about having Rosie handle tagging a pull for extended tests and a regular nightly run?

jhadvig commented:
@deads2k not yet, but ocf thats the next step

smarterclayton commented:
1529f489adcef368746b20c88a3c83f17bdce5d3 is breaking running hack/test-cmd.sh on Mac (requires a password) and on travis.  Reverting it for now.

smarterclayton commented:
I reverted the hack/test-cmd changes, please open a follow up and verify that it works without access to sudo
